pkg/datapath/linux/devices_controller.go

```diff
@@ -345,6 +352,21 @@ func (dc *devicesController) initialize() error {
 
 	txn.Commit()
 
+	// Remove non-existent devices in the device bpf map
+	dc.deviceMap.IterateWithCallback(
+		func(key *devicemap.DeviceKey, val *devicemap.DeviceValue) {
+			d, _, _ := dc.params.DeviceTable.Get(txn, tables.DeviceIDIndex.Query(int(key.IfIndex)))
+			if d == nil {
+				err := dc.deviceMap.Delete(key.IfIndex)
+				if err != nil {
+					dc.log.Warn("Failed to delete entry from device bpf map",
+						logfields.Error, err,
+						logfields.Device, d,
+					)
+				}
+			}
+		})
+
```

Comment:
We shouldn't manage the device map directly from this device controller. The whole point of this logic is to populate the StateDB tables which then can be consumed in a decoupled way.

This logic should be moved to https://github.com/cilium/cilium/blob/main/pkg/datapath/orchestrator/orchestrator.go

It is the job of the orchestrator to initialize the datapath on startup and to re-initialize when something like the set of current devices changes (it consumes the devices StateDB table for this purpose).

It should consume the table and reconcile the map as part of the (re)initialization process.

Reply:
This logic has been moved to Reinitialize method in the loader module. Please take a look, Thanks.